agent/internal/docker.go

```diff
@@ -304,18 +304,175 @@ func (d *dockerActor) sendAuxLog(ctx *actor.Context, msg string) {
+//
+// Showing a per-layer status bar is practically impossible without an interactive terminal (as
+// docker run would have).
+//
+// So instead we create a weighted-average status bar, where every layer's download and extraction
+// count as equal parts.  The status bar ends up pretty jerky but it still gives a "sensation" of
+// progress; things don't look frozen, the user has a rough idea of how far along you are, and the
+// logs are still sane afterwards.
+func (f *pullLogFormatter) RenderProgress() string {
+	var downloaded int64
+	var extracted int64
+	progress := 0.0
+	for _, id := range f.Order {
+		info := f.Known[id]
+		downloaded += info.Downloaded
+		extracted += info.Extracted
+		switch {
+		case !info.DownloadStarted:
+			// no progress on this layer
+		case info.Extracted == info.Total:
+			// this layer is complete
+			progress += 1.0
+		case info.Downloaded == info.Total:
+			// download complete, extraction in progress
+			progress += 0.5 + 0.5*float64(info.Extracted)/float64(info.Total)
+		default:
+			progress += 0.5 * float64(info.Extracted) / float64(info.Total)
+		}
+	}
+
+	// normalize by layer count
+	progress /= float64(len(f.Known))
+
+	// 40-character progress bar
+	prog := int(40.0 * progress)
+
+	bar := ""
+	for i := 0; i < 40; i++ {
+		if prog > i {
+			bar += "="
+		} else {
+			bar += " "
+		}
+	}
+
+	return fmt.Sprintf(
+		"[%v] Downloaded: %.1fMB, Extracted %.1fMB",
+		bar,
+		float64(downloaded)/1e6,
+		float64(extracted)/1e6,
+	)
+}
+
+func (f *pullLogFormatter) backoffOrRenderProgress() *string {
+	// log at most one line every 1 second
+	now := time.Now()
+	if now.Before(f.Backoff) {
+		return nil
+	}
+	f.Backoff = now.Add(1 * time.Second)
+
+	progress := f.RenderProgress()
+	return &progress
+}
+
+// Update returns nil or a rendered progress update for the end user.
+func (f *pullLogFormatter) Update(msg jsonmessage.JSONMessage) *string {
+	if msg.Error != nil {
+		fmt.Printf("%v\n", msg.Error)
+		return nil
+	}
+
+	if msg.Status == "Pulling fs layer" || msg.Status == "Waiting" {
+		if _, ok := f.Known[msg.ID]; !ok {
+			// New layer!
+			f.Known[msg.ID] = &pullInfo{}
+			f.Order = append(f.Order, msg.ID)
+		}
+		return nil
+	}
+
+	if msg.Status == "Downloading" {
+		info := f.Known[msg.ID]
+		if info.ExtractStarted {
+			panic("already extracting!")
+		}
+		info.Downloaded = msg.Progress.Current
+		// The first "Downloading" msg is important, as it gives us the layer size.
+		if !info.DownloadStarted {
+			info.DownloadStarted = true
+			info.Total = msg.Progress.Total
+		}
+		return f.backoffOrRenderProgress()
+	}
+
+	if msg.Status == "Extracting" {
+		info := f.Known[msg.ID]
+		info.Extracted = msg.Progress.Current
+		if !info.ExtractStarted {
+			info.ExtractStarted = true
+			// Forcibly mark Downloaded as completed.
+			info.Downloaded = info.Total
+		}
+		return f.backoffOrRenderProgress()
+	}
+
+	if msg.Status == "Pull complete" {
```

Comment:
are these statuses available as consts somewhere within docker (don't want to break if they change them)

Reply:
also, i'd just switch on status, maybe emit a failure log if no cases match

Reply:
Yeah, it would be insane if they were just hardcoded strings on something that seems to looks like an API, right?

...

"Downloading" [here](https://github.com/moby/moby/blob/master/plugin/fetch_linux.go#L280), [here](https://github.com/moby/moby/blob/master/distribution/pull_v2.go#L227)
"Extracting" [here](https://github.com/moby/moby/blob/master/distribution/xfer/download.go#L346), [here](https://github.com/moby/moby/blob/master/plugin/fetch_linux.go#L140)
"Pull complete" [here](https://github.com/moby/moby/blob/master/distribution/xfer/download.go#L374)

Re: emit a failure log if no cases match: I'd prefer to capture what's necessary and ignore the other (not particularly meaningful) messages.  It's not really structured well enough to deal with it any other way.